Draft the "Patterns of collaboration" section (#15)
DavisVaughan/code-review: collaboration/index.qmd

# Patterns of collaboration {#sec-patterns-of-collaboration .unnumbered}

Within the tidyteam group, the way we collaborate with colleagues and external contributors is highly dependent on the context of who is working on a particular package, how closely the contributors are working together, and how often that package is updated.
A large amount of the advice in this guide is independent of any specific collaboration context, but some aspects, such as the expected turnaround time of a code review, and the level of detail that is expected from a review, is closely tied to it.

It is important to keep in mind that package development at Posit is probably a little different from software development at other companies.

> Packages often lie dormant for awhile as the core maintainer works on other projects, which means that sometimes PRs take longer to get reviewed than expected.

Reviewer: This feels specific to the tidy teams rather than Posit?

### The understudy {#sec-the-understudy}

An understudy typically follows a project from afar, generally submitting smaller PRs or doing superficial code review with a focus on API design, unit tests, and documentation.
This allows the understudy to stay up to date with a project they may potentially take over in the future.

When an understudy submits a PR, review is typically swift, in-depth, and comes from a core maintainer.
Understudy PRs are often smaller in scope, and can occasionally require multiple rounds of review if they aren't intimately familiar with the specifics of the package they are working on.
If both the understudy and reviewer are Posit employees, discussing the review live (i.e. over Zoom) is often very beneficial for the understudy's overall understanding of the package.

When an understudy reviews a PR, review is expected to be swift, and the PR assigned to the understudy for review should not require an in-depth knowledge of the package.
If the understudy isn't available to review, it is fine to merge the PR without waiting for a review, especially if it is blocking some other work.

> An understudy may take the form of an intern at Posit, a motivated community member, or another tidyteam member.

Reviewer: ```suggestion
An understudy may be an intern at Posit, a motivated community member, or another tidyteam member.
```

### External contributions {#sec-external-contributions}

An external contribution is a PR authored by someone who isn't a regular contributor to a package.
This may be a one-off PR from a community member, but is often another tidyteam member that notices a small bug or feature that is missing from a package they don't typically work on.
For example, I am not a regular contributor to usethis, but noticed a small non-critical bug and [sent in a PR for it](https://github.com/r-lib/usethis/pull/1786).

When an external contributor authors a PR, they should not expect an immediate review.

> If the package is in a dormant state, it is possible that the PR won't be looked at for weeks at a time.

Reviewer: ```suggestion
If the package is in a dormant state, it is possible that the PR won't be looked at for weeks or even months at a time.
```
